contracts/sol6/reserves/bridgeReserve/uniswap/KyberUniswapReserve.sol

```diff
@@ -0,0 +1,397 @@
+
+        delete tokenExchange[address(token)];
+        
+        token.safeApprove(uniswapPair, 0);
+
+        TokenDelisted(token);
+    }
+
+    function isValidTokens(
+        IERC20 src,
+        IERC20 dest
+    )
+        public
+        view
+        returns(bool)
+    {
+        return (
+            (
+                src == ETH_TOKEN_ADDRESS &&
+                tokenExchange[address(dest)] != address(0)
+            ) ||
+            (
+                tokenExchange[address(src)] != address(0) &&
+                dest == ETH_TOKEN_ADDRESS
+            )
+        );
+    }
+
+    event TradeEnabled(
+        bool enable
+    );
+
+    function enableTrade()
+        external
+        returns(bool)
+    {
+        onlyAdmin();
+        tradeEnabled = true;
+        TradeEnabled(true);
+        return true;
+    }
+
+    function disableTrade()
+        external
+        returns(bool)
+    {
+        onlyAlerter();
+        tradeEnabled = false;
+        TradeEnabled(false);
+        return true;
+    }
+
+    event KyberNetworkSet(
+        address kyberNetwork
+    );
+
+    function setKyberNetwork(
+        address _kyberNetwork
+    )
+        public
+    {
+        onlyAdmin();
+        require(_kyberNetwork != address(0));
+        if (kyberNetwork != _kyberNetwork){
+            kyberNetwork = _kyberNetwork;
+            KyberNetworkSet(kyberNetwork);
+        }
+    }
+
+    function deductFee(
+        uint amount
+    )
+        public
+        view
+        returns(uint)
+    {
+        return amount * (10000 - feeBps) / 10000;
+    }
+
+
+    function calcUniswapConversion(
+        IERC20 src,
+        IERC20 dest,
+        uint srcQty
+    )
+        internal
+        view
+        returns(uint destQty, uint rate)
+    {
+        IUniswapV2Pair uniswapPair;
+        if (src == ETH_TOKEN_ADDRESS) {
+            uniswapPair = IUniswapV2Pair(tokenExchange[address(dest)]);
+
+            uint amountLessFee = deductFee(srcQty);
+            if (amountLessFee == 0) return (0, 0);
+
+
+            (uint reserveIn, uint reserveOut) = getReserves(uniswapPair, WETH, address(dest));
+
+            destQty = UniswapV2Library.getAmountOut(amountLessFee, reserveIn, reserveOut);
+        } else {
+            uniswapPair = IUniswapV2Pair(tokenExchange[address(src)]);
+            (uint reserveIn, uint reserveOut) = getReserves(uniswapPair, address(src), WETH);
+            destQty = deductFee(
+                UniswapV2Library.getAmountOut(srcQty, reserveIn, reserveOut)
+            );
+        }
+
+        rate = calcRateFromQty(
+            srcQty, /* srcAmount */
+            destQty, /* destAmount */
+            getDecimals(src), /* srcDecimals */
+            getDecimals(dest) /* dstDecimals */
+        );
+    }
+
+    function getReserves(IUniswapV2Pair uniswapPair, address tokenA, address tokenB) internal view returns (uint reserveA, uint reserveB) {
+        (address token0,) = UniswapV2Library.sortTokens(tokenA, tokenB);
```

Comment:
can we have sorted tokens stored once per token listing.
And avoid extra call to sort token?

Reply:
but actually not sure its less gas. depends if sort token is external call or not.

Reply:
This is internal library call. So it does not cost much